mbed-drivers/v1/I2CDetail.hpp

```diff
@@ -0,0 +1,338 @@
+    /**
+     * @brief Starts the transaction at the head of the queue
+     */
+    virtual I2CError start_transaction() = 0;
+
+    /**
+     * @brief Starts the next segment
+     */
+    virtual I2CError start_segment() = 0;
+
+    /**
+     * @brief Validates the transaction according to the criteria of the derived Resource Manager
+     * @param[in] transaction the transaction to validate
+     */
+    virtual I2CError validate_transaction(I2CTransaction *transaction) const = 0;
+
+protected:
+    /**
+     * @brief Process an event
+     *
+     * Handles the mechanics of processing an event, including
+     * * handling errors
+     * * Triggering any irq callback
+     * * (optionally) scheduling an event handler using handle_event
+     * * (optionally) advancing to the next transaction
+     * * starting the next segment
+     *
+     * @param[in] event the source of the current handler call
+     */
+    void process_event(uint32_t event);
+
+    /**
+     * @brief Handle an event
+     *
+     * Distributes the event and transaction that triggered it to any event handler with a matching event mask, then
+     * deletes the transaction, using the method specified by the issuing I2C object
+     *
+     * @param[in] t the transaction that was in progress when the event was triggered
+     * @param[in] event the event(s) that occurred
+     */
+    void handle_event(I2CTransaction *t, uint32_t event);
+
+    /**
+     * @brief The resource manager constructor.
+     *
+     * Resource managers may not be copied or moved. The resource manager must be inherited.
+     */
+    I2CResourceManager();
+
+    /**
+     * The I2CResourceManager destructor. It is not expected that the destructor will be called, however, the destructor
+     * handles cleanup of resources such as deleting any pending transactions.
+     */
+    ~I2CResourceManager();
+
+    // The head of the transaction queue
+    I2CTransaction * volatile TransactionQueue;
+};
+
+I2CResourceManager * get_i2c_owner(int I);
+
+#define I2C_TRANSACTION_NHANDLERS 4
+
+/**
+ * @brief A helper class for holding a callback and an event mask
+ */
+class I2CEventHandler {
+public:
+    /**
+     * @brief Constructor for I2CEventHandler
+     *
+     * Sets the event mask to 0 and the callback to null
+     */
+    I2CEventHandler();
+
+    /**
+     * @brief Handle an event
+     *
+     * If event & _event_mask is non-zero, calls the event handler, otherwise does nothing
+     *
+     * @param[in] t the transaction that was in progress when the event was triggered
+     * @param[in] event the event(s) that occurred
+     */
+    void call(I2CTransaction *t, uint32_t event);
+
+    /**
+     * @brief Set the callback and the event that event mask
+     *
+     * @param[in] cb the callback to trigger when the event mask is matched
+     * @param[in] event The event mask to use to filter events
+     */
+    void set(const I2C_event_callback_t &cb, uint32_t event);
```

Comment:
They are all typedefed from eachother. The problem is include order. This is in the detail directory, which means that `event_callback_t` has not been declared yet. This wouldn't be a problem if I could place `event_callback_t` at global scope, but that would result in namespace collisions.  This would be resolved by rfc#14